### ISSUE TITLE
Async version of applyFilters

### ISSUE THREAD

**Issue by Tofandel** [WordPress/gutenberg#35509]
## What problem does this address?

The js hooks follow the php logic and currently don't allow for async actions to be performed and return a result

## What is your proposed solution?

Javascript has some very different concepts than php and it would be nice to have `applyFiltersAsync` returning chained promises to be able to fully use javascript's features and to do for example

```js
addFilter('some_filter', 'namespace', async (res) => {
  // some async operations
  return { res, ...await fetch('something') };
})
addFilter('some_filter', 'namespace', async (res) => {
  // should wait for previous filter
  // some async operations
  return { res, ...await fetch('some other thing') };
})
const res = await applyFiltersAsync('some_filter', 'namespace', {})
```

**Comment by blakewilson** [comment]
Any movement on this? This would be immensely helpful

**Comment by Tofandel** [comment]
I think you can already send a promise to applyFilter and use that from the callback, it should achieve the same thing with more control on which filter to wait for

Eg
```js
addFilter('some_filter', 'namespace', async (chain) => {
  // some async operations
  return { ...await chain, ...await fetch('something') };
})
addFilter('some_filter', 'namespace', (chain) => {
  // Run this in parallel without waiting
  return Promise.all([chain, fetch('some other thing')]);
})
const res = await applyFilters('some_filter', 'namespace', Promise.resolve())
```
